File: backend/worker.py
"""
Simplified Cloudflare Worker for Figma Product Catalog API
Using built-in Python modules only
"""
from workers import WorkerEntrypoint, Response
import json


class Default(WorkerEntrypoint):
    async def fetch(self, request):
        """Handle incoming requests"""
        url = request.url
        method = request.method
        path = url.split('?')[0].split('/')[-1]

        # Basic routing
        if method == "GET":
            if path == "" or path == "api":
                return self.handle_root()
            elif path == "products":
                return self.handle_get_products()
            elif path == "health":
                return self.handle_health()
        elif method == "POST":
            if path == "products":
                return await self.handle_create_product(request)
        elif method == "OPTIONS":
            # Handle CORS preflight
            return self.handle_cors_preflight()

        # 404 for unknown routes
        return Response(
            json.dumps({"error": "Not found"}),
            status=404,
            headers=self.get_cors_headers()
        )

    def get_cors_headers(self):
        """Get CORS headers for responses"""
        return {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "GET, POST, PUT, DELETE, OPTIONS",
            "Access-Control-Allow-Headers": "*"
        }

    def handle_cors_preflight(self):
        """Handle CORS preflight requests"""
        return Response(
            "",
            status=204,
            headers=self.get_cors_headers()
        )

    def handle_root(self):
        """Root endpoint"""
        response_data = {
            "message": "Figma Product Catalog API on Cloudflare Workers",
            "version": self.env.API_VERSION,
            "endpoints": {
                "/api": "API info",
                "/api/health": "Health check",
                "/api/products": "Products endpoint (GET/POST)"
            }
        }
        return Response(
            json.dumps(response_data),
            headers=self.get_cors_headers()
        )

    def handle_health(self):
        """Health check endpoint"""
        return Response(
            json.dumps({"status": "healthy", "service": "Cloudflare Workers Python"}),
            headers=self.get_cors_headers()
        )

    def handle_get_products(self):
        """Get products - example data"""
        products = [
            {
                "id": 1,
                "name": "Букет роз",
                "price": 15000,
                "currency": "KZT",
                "available": True,
                "category": "bouquets",
                "image": "https://example.com/roses.jpg"
            },
            {
                "id": 2,
                "name": "Тюльпаны",
                "price": 8000,
                "currency": "KZT",
                "available": True,
                "category": "flowers",
                "image": "https://example.com/tulips.jpg"
            },
            {
                "id": 3,
                "name": "Орхидея в горшке",
                "price": 25000,
                "currency": "KZT",
                "available": False,
                "category": "potted",
                "image": "https://example.com/orchid.jpg"
            },
            {
                "id": 4,
                "name": "Свадебный букет",
                "price": 45000,
                "currency": "KZT",
                "available": True,
                "category": "wedding",
                "image": "https://example.com/wedding.jpg"
            }
        ]

        return Response(
            json.dumps({"products": products, "total": len(products)}),
            headers=self.get_cors_headers()
        )

    async def handle_create_product(self, request):
        """Create a new product (example - doesn't persist)"""
        try:
            # Get request body
            body = await request.text()
            product_data = json.loads(body)

            # Add ID and return
            product_data["id"] = 999
            product_data["created_at"] = "2025-09-26T18:00:00Z"

            return Response(
                json.dumps({
                    "message": "Product created successfully",
                    "product": product_data
                }),
                status=201,
                headers=self.get_cors_headers()
            )
        except Exception as e:
            return Response(
                json.dumps({"error": str(e)}),
                status=400,
                headers=self.get_cors_headers()
            )

File: crm-bitrix/src/worker.js
const BITRIX_API_URL = 'https://cvety.kz/api/v2';
const BITRIX_TOKEN = 'ABE7142D-D8AB-76AF-8D6C-2C4FAEA9B144';

export default {
  async fetch(request, env) {
    const url = new URL(request.url);
    const pathname = url.pathname;

    // Proxy API requests through Cloudflare to avoid CORS issues
    if (pathname.startsWith('/api/v2/')) {
      const apiPath = pathname.replace('/api/v2', '');
      const apiUrl = new URL(apiPath + url.search, BITRIX_API_URL);

      const apiRequest = new Request(apiUrl, {
        method: request.method,
        headers: {
          'Authorization': `Bearer ${BITRIX_TOKEN}`,
          'Content-Type': 'application/json',
        },
        body: request.method === 'GET' ? undefined : await request.text(),
      });

      try {
        const response = await fetch(apiRequest);

        // Clone response and add CORS headers
        const corsResponse = new Response(response.body, {
          status: response.status,
          statusText: response.statusText,
          headers: new Headers(response.headers),
        });

        // Set proper CORS headers (single origin only)
        corsResponse.headers.set('Access-Control-Allow-Origin', '*');
        corsResponse.headers.set('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS, PATCH');
        corsResponse.headers.set('Access-Control-Allow-Headers', 'Content-Type, Authorization');
        corsResponse.headers.set('Access-Control-Max-Age', '3600');

        // Remove duplicate CORS headers if present
        corsResponse.headers.delete('access-control-allow-origin');
        corsResponse.headers.set('Access-Control-Allow-Origin', '*');

        return corsResponse;
      } catch (error) {
        return new Response(
          JSON.stringify({ success: false, error: 'API request failed: ' + error.message }),
          {
            status: 500,
            headers: {
              'Content-Type': 'application/json',
              'Access-Control-Allow-Origin': '*',
            }
          }
        );
      }
    }

    // Handle OPTIONS preflight requests
    if (request.method === 'OPTIONS') {
      return new Response(null, {
        status: 204,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS, PATCH',
          'Access-Control-Allow-Headers': 'Content-Type, Authorization',
          'Access-Control-Max-Age': '3600',
        },
      });
    }

    try {
      // Try to serve the requested file from assets
      let response = await env.ASSETS.fetch(request);

      // If not found and it's a route (not an asset), serve index.html for SPA
      if (response.status === 404 && !pathname.includes('.')) {
        response = await env.ASSETS.fetch(new Request(new URL('/index.html', url).toString()));
      }

      return response;
    } catch (error) {
      // Fallback: serve index.html for SPA routing
      return env.ASSETS.fetch(new Request(new URL('/index.html', url).toString()));
    }
  },
};
